man/foreign.doc

```diff
@@ -951,6 +969,42 @@ advised to use the *_chars() or *_nchars() functions whenever
 applicable.
 
 
+\subsection{Input and output}
+\label{sec:input-and-output}
+
+For input and output, \file{SWI-Stream.h} defines a
+set of functions that are similar to the C library functions, except
+prefixed by \textbf{S}, e.g. Sfprintf(). They differ from the C
+functions in following ways:
+\begin{itemize}
+  \item Instead of returning the number of bytes written and a negative
+    value for error, they return \const{TRUE} and \const{FALSE}.
+  \item Instead of a \ctype{FILE}, they access the Prolog streams,
+    using \ctype{IOSTREAM*}. In particular, \const{Scurrent_output}
+    accesses the current output stream (similarly,
+    \const{Scurrent_intput}, \const{Suser_output},
+    \const{Suser_error}, \const{Suser_input}).
+  \item If you wish to directly use the operating system's
+    \const{stdin}, \const{stdout}, \const{stderr}, you can use
+    \const{Sinput}, \const{Soutput}, \const{Serror}.
+\end{itemize}
+
+In general, if a stream is acquired via PL_acquire_stream(), an error
+is raised when PL_release_stream() is called, so in that situation,
+there's no need to check the return codes from the IO functions. Blob
+write callbacks are also called in the context of an acquired stream,
+so there is no need to check the return codes from its IO function
+calls.  However, if you use one of the standard streams such as
```

Comment:
Is  this correct? I've added checks for return codes in the example code snippets, on the assumption that an IO error also sets qf->exception or similar.

Reply:
It doesn't set qf->exception.   PL_release_stream() should set the default context exception, which means it binds a global internal `term_t` reserved for exceptions.    The qf->exception is to deal with exceptions when using PL_open_query() and friends.

Reply:
But what about errors happening on `Suser_output` etc.? Should the programmer do `PL_acquire_stream(Suser_output)` in the foreign function? That might be good practice anyway, to ensure the stream is locked so that multithreading doesn't interleave outputs.

(Which reminds me: I'll add an RAII version of PL_acquire_stream())